src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,192 @@
+            if (Controller.IsFilterCharacter(roslynItem, typeChar, filterText))
+            { 
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            var serviceRules = completionService.GetRules();
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+
+            // Tab, Enter and Null (call invoke commit) are always a commit character. 
+            if (typeChar != '\t' && typeChar != '\n' && typeChar != '\0' && !Controller.IsCommitCharacter(serviceRules, roslynItem, typeChar, filterText))
+            {
+                return new AsyncCompletionData.CommitResult(isHandled: true, AsyncCompletionData.CommitBehavior.CancelCommit);
+            }
+
+            if (!item.Properties.TryGetProperty<ITextSnapshot>(CompletionSource.TriggerSnapshot, out var triggerSnapshot))
+            {
+                return CommitResultUnhandled;
+            }
+
+            // Commit with completion serivce assumes that null is provided is case of invoke. VS provides '\0' in the case.
+            char? commitChar = typeChar == '\0' ? null : (char?)typeChar;
+            var commitBehavior = Commit(
+                document, completionService, session.TextView, subjectBuffer, 
+                roslynItem, commitChar, triggerSnapshot, serviceRules, filterText, cancellationToken);
+
+            return new AsyncCompletionData.CommitResult(isHandled: true, commitBehavior);            
+        }
+
+        private AsyncCompletionData.CommitBehavior Commit(
+            Document document,
+            CompletionService completionService,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            char? commitCharacter,
+            ITextSnapshot triggerSnapshot,
+            CompletionRules rules,
+            string filterText,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer did not provide Edit Access"));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                FatalError.ReportWithoutCrash(new InvalidOperationException("Subject buffer is editing by someone else."));
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
```

Comment:
why are we not using "CaretPreservingEditTransaction" ?

Reply:
Yes, the code is here: [https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FLanguage%2FImpl%2FLanguage%2FAsyncCompletion&version=GBmaster](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FLanguage%2FImpl%2FLanguage%2FAsyncCompletion&version=GBmaster)

@jinujoseph to verify if this code can be seen out of DevDiv.

---
In reply to: [232429871](https://github.com/dotnet/roslyn/pull/29016#discussion_r232429871) [](ancestors = 232429871)